File: julia/src/StarPU.jl
# StarPU --- Runtime system for heterogeneous multicore architectures.
#
# Copyright (C) 2020-2021  Université de Bordeaux, CNRS (LaBRI UMR 5800), Inria
#
# StarPU is free software; you can redistribute it and/or modify
# it under the terms of the GNU Lesser General Public License as published by
# the Free Software Foundation; either version 2.1 of the License, or (at
# your option) any later version.
#
# StarPU is distributed in the hope that it will be useful, but
# WITHOUT ANY WARRANTY; without even the implied warranty of
# MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.
#
# See the GNU Lesser General Public License in COPYING.LGPL for more details.
#
"""
__precompile__()
"""
module StarPU
import Libdl
using CBinding

include("utils.jl")

const starpu_wrapper_library_name=fstarpu_task_library_name()

include("translate_headers.jl")

if !isfile(joinpath(fstarpu_build_dir(), "julia/gen/libstarpu_common.jl")) || !isfile(joinpath(fstarpu_build_dir(), "julia/gen/libstarpu_api.jl")) ||
    mtime(joinpath(@__FILE__, "translate_headers.jl")) > mtime(joinpath(fstarpu_build_dir(), "julia/gen/libstarpu_api.jl"))
    starpu_translate_headers()
end

include(joinpath(fstarpu_build_dir(), "julia/gen/libstarpu_common.jl"))
include(joinpath(fstarpu_build_dir(), "julia/gen/libstarpu_api.jl"))
include("globals.jl")

include("compiler/include.jl")
include("linked_list.jl")
include("destructible.jl")
include("perfmodel.jl")
include("data.jl")
include("blas.jl")
include("task.jl")
include("task_dep.jl")
include("init.jl")

# macro
export @starpu_filter
export @starpu_block
export @starpu_async_cl
export @starpu_sync_tasks

# enum / define
export STARPU_CPU
export STARPU_CUDA
export STARPU_CUDA_ASYNC
export STARPU_OPENCL
export STARPU_MAIN_RAM
export StarpuDataFilterFunc
export STARPU_MATRIX_FILTER_VERTICAL_BLOCK, STARPU_MATRIX_FILTER_BLOCK
export STARPU_VECTOR_FILTER_BLOCK
export STARPU_PERFMODEL_INVALID, STARPU_PER_ARCH, STARPU_COMMON
export STARPU_HISTORY_BASED, STARPU_REGRESSION_BASED
export STARPU_NL_REGRESSION_BASED, STARPU_MULTIPLE_REGRESSION_BASED
export starpu_tag_t
export STARPU_NONE,STARPU_R,STARPU_W,STARPU_RW, STARPU_SCRATCH
export STARPU_MPI_REDUX, STARPU_REDUX,STARPU_COMMUTE, STARPU_SSEND, STARPU_LOCALITY
export STARPU_ACCESS_MODE_MAX

# BLAS
export STARPU_SAXPY

# functions
export starpu_cublas_init
export starpu_init
export starpu_shutdown
export starpu_memory_pin
export starpu_memory_unpin
export starpu_data_access_mode
export starpu_data_acquire_on_node
export starpu_data_release_on_node
export starpu_data_unregister
export starpu_data_register
export starpu_data_get_sub_data
export starpu_data_partition
export starpu_data_unpartition
export starpu_data_map_filters
export starpu_data_wont_use
export starpu_task_insert
export starpu_task_wait_for_all
export starpu_task_submit
export starpu_task_end_dep_add
export starpu_task_end_dep_release
export starpu_task_declare_deps
export starpu_task_declare_end_deps
export starpu_task_wait_for_n_submitted
export starpu_task_destroy
export starpu_tag_remove
export starpu_tag_wait
export starpu_tag_notify_from_apps
export starpu_iteration_pop
export starpu_iteration_push
export starpu_tag_declare_deps
export starpu_task
export starpu_task_wait
export starpu_codelet
export starpu_perfmodel
export starpu_perfmodel_type
export starpu_translate_headers
export starpu_data_get_default_sequential_consistency_flag
export starpu_data_set_default_sequential_consistency_flag
export starpu_data_get_sequential_consistency_flag
export starpu_data_set_sequential_consistency_flag
export starpu_worker_get_count
export starpu_cpu_worker_get_count
export starpu_cuda_worker_get_count
export starpu_opencl_worker_get_count

end

File: starpujni/src/main/java/fr/labri/hpccloud/starpu/StarPU.java
// StarPU --- Runtime system for heterogeneous multicore architectures.
//
// Copyright (C) 2020-2022  Université de Bordeaux, CNRS (LaBRI UMR 5800), Inria
//
// StarPU is free software; you can redistribute it and/or modify
// it under the terms of the GNU Lesser General Public License as published by
// the Free Software Foundation; either version 2.1 of the License, or (at
// your option) any later version.
//
// StarPU is distributed in the hope that it will be useful, but
// WITHOUT ANY WARRANTY; without even the implied warranty of
// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.
//
// See the GNU Lesser General Public License in COPYING.LGPL for more details.
//
package fr.labri.hpccloud.starpu;

import fr.labri.hpccloud.starpu.data.DataHandle;

import java.io.IOException;
import java.net.URL;
import java.util.Map;
import java.util.Properties;

public class StarPU
{
	public static boolean enableTrace = false;

	static
	{
		try
		{
			String libpath = NativeLibInfo.PATH_TO_LIB;
			System.err.println("Try to load " + libpath);
			NativeUtils.loadLibraryFromJar(libpath);
		}
		catch (IOException | NullPointerException e)
		{
			e.printStackTrace();
		}
	}

	public static void setenv(String variable, String value, boolean overwrite)
	{
		System.out.println("setenv(" + variable + "," + value);
		setenv_(variable, value, overwrite);
	}

	private static native void setenv_(String variable, String value, boolean overwrite);

	public static void init() throws StarPUException
	{
		init_();
	}

	public static void init(Map<String, String> env) throws StarPUException
	{
		for (Map.Entry<String, String> e : env.entrySet())
		{
			setenv(e.getKey(), e.getValue(), true);
		}
		init_();
	}

	private static native void init_() throws StarPUException;

	public static void shutdown() throws StarPUException
	{
		shutdown_();
	}

	private static native void shutdown_() throws StarPUException;

	public static void submitTask(Codelet codelet, boolean synchronous, DataHandle... handles) throws StarPUException
	{
		long taskID = submitTask_(codelet, handles);
		if (synchronous)
		{
			waitForTasks(new long[]{taskID});
		}
	}

	public static native long submitTask_(Codelet codelet, DataHandle... handles) throws StarPUException;

	public static native void waitForTasks(long tasks[]) throws StarPUException;

	public static void mapCodelet(Codelet mapCl, boolean waitForAll, DataHandle input, DataHandle output) throws StarPUException
	{
		int nbChildren = input.getNbChildren();

		if (nbChildren == 0)
		{
			submitTask(mapCl, waitForAll, input, output);
		}
		else
		{
			long taskIDs[] = new long[nbChildren];

			DataHandle[] subHandles = new DataHandle[2];
			for (int i = 0; i < nbChildren; i++)
			{
				subHandles[0] = input.getSubData(i);
				subHandles[1] = output.getSubData(i);
				taskIDs[i] = submitTask_(mapCl, subHandles);
			}
			if (waitForAll)
			{
				waitForTasks(taskIDs);
			}
		}
	}

	public static void joinCodelet(Codelet joinCl, boolean waitForAll, DataHandle input1, DataHandle input2, DataHandle output) throws StarPUException
	{
		int nbOutHdl = output.getNbChildren();
		assert (nbOutHdl == input1.getNbChildren());
		int nbTasks = (input1.getNbHandles() * input2.getNbHandles());
		long taskIDs[] = new long[nbTasks];
		int i = 0;
		DataHandle[] subHandles = new DataHandle[3];

		if (input1.getNbChildren() == 0)
		{
			subHandles[0] = input1;
			subHandles[2] = output;
			for (DataHandle hdl2 : input2.getHandles())
			{
				subHandles[1] = hdl2;
				taskIDs[i] = submitTask_(joinCl, subHandles);
				i++;
			}
		}
		else
		{
			for (int h = 0; h < nbOutHdl; h++)
			{
				subHandles[0] = input1.getSubData(h);
				subHandles[2] = output.getSubData(h);
				for (DataHandle hdl2 : input2.getHandles())
				{
					subHandles[1] = hdl2;
					taskIDs[i] = submitTask_(joinCl, subHandles);
					i++;
				}
				subHandles[2] = output;
			}
		}
		if (waitForAll)
			waitForTasks(taskIDs);
	}

	public static void reduceCodelet(Codelet reduceCl, boolean waitForAll, DataHandle input, DataHandle accumulator) throws StarPUException
	{
		if (input.getNbChildren() > 0)
		{
			input.unpartition();
		}
		submitTask(reduceCl, waitForAll, input, accumulator);
	}

	public static native void taskWaitForAll() throws StarPUException;

	public static native double drand48();

	public static native boolean runNativeTests();
}
